File: CivitasJava/src/civitas/Jugador.java
package civitas;

import GUI.*;
import java.util.*;

public class Jugador implements Comparable<Jugador>{
    protected static int CasasMax=4;
    protected static int CasasPorHotel=4;
    protected boolean encarcelado;
    protected static int HotelesMax=4;
    private String nombre;
    private int numCasillaActual;
    protected static float PasoPorSalida=1000;
    protected static float PrecioLibertad=200;
    boolean puedeComprar;
    private float saldo;
    private static float SaldoInicial=7500;
    SorpresaSalirCarcel salvoconducto;
    ArrayList<TituloPropiedad> arrayPropiedades;
    boolean especulador;
            
    boolean vender(int ip){
        boolean done = false;
        if(!encarcelado){
            if(existeLaPropiedad(ip)){
                boolean done2 = arrayPropiedades.get(ip).vender(this);
                if(done2){
                    arrayPropiedades.remove(ip);
                    Diario.getInstance().ocurreEvento("Se vende propiedad ");
                    done=true;
                }
            }  
        }
        
        return done;
    }
    
    public String toString(){
        String mensaje="******************************\n";
        mensaje += nombre + "\n******************************\n\n"; 
        mensaje += "ENCARCELADO = " + encarcelado; 
        mensaje += "\n Numero Casilla Actual: " + numCasillaActual;   
        mensaje += "\n Saldo: " + saldo;
        mensaje += "\n Puede comprar = " + puedeComprar;
        mensaje += "\n CasasMax= " + CasasMax;
        mensaje += "\n Casas por hotel= " + CasasPorHotel;
        mensaje += "\n Hoteles max= " + HotelesMax;
        mensaje += "\n Paso por salida= " + PasoPorSalida;
        mensaje += "\n Precio Libertad = " + PrecioLibertad;
        mensaje += "\n Saldo inicial = " + SaldoInicial;
        
        return mensaje;
    }
    
    boolean tieneSalvoconducto(){
        return (salvoconducto != null);
    }
    
    boolean tieneAlgoQueGestionar(){
        return arrayPropiedades.size() > 0;
    }
    
    boolean salirCarcelTirando(){
        boolean done = false;
        if(encarcelado && Dado.getInstance().salgoDeLaCarcel()){
            done=true;
            Diario.getInstance().ocurreEvento("Jugador sale de la carcel tirando");
            encarcelado=false;
        }
        
        return done;
    }
    
    boolean salirCarcelPagando(){
        boolean done = false;
        if(encarcelado && puedeSalirCarcelPagando()){
            paga(PrecioLibertad);
            encarcelado=false;
            done=true;
            Diario.getInstance().ocurreEvento("Jugador sale de la carcel pagando");
        }
        
        return done;
    }
    
    boolean recibe(float cantidad){
        boolean done=false;
        if(!encarcelado){
            modificarSaldo(cantidad);
            done=true;
        }
        
        return done;
    }
    
    boolean puedoGastar(float precio){
        boolean puedo=(saldo >= precio);
        
        return puedo;
    }
    
    boolean puedoEdificarHotel(TituloPropiedad propiedad){
        boolean done=false;
        float precio=propiedad.getPrecioEdificar();
        if(puedoGastar(precio)){
            if(propiedad.getNumHoteles()<HotelesMax){
                if(propiedad.getNumCasas()>=CasasPorHotel){
                    done=true;
                }
            }
        }
        
        return done;
    }
    
    boolean puedoEdificarCasa(TituloPropiedad propiedad){
        boolean done=false;
        float precio=propiedad.getPrecioEdificar();
        if(puedoGastar(precio)){
            if(propiedad.getNumCasas()<CasasMax){
                done=true;
            }
        }
        
        return done;
    }
    
    private boolean puedeSalirCarcelPagando(){
        return puedoGastar(PrecioLibertad);
    }
    
    public boolean puedeComprarCasilla(){
        boolean puede=false;
        if(!encarcelado){
            if(puedeComprar){
                puede=true;
            }
        }
        
        return puede;
    }
    
    void perderSalvoconducto(){
        Diario.getInstance().ocurreEvento("Jugador " + getNombre() + " pierde salvoconducto");
        salvoconducto.usada();
        salvoconducto=null;
    }
    
    boolean pasaPorSalida(){
        modificarSaldo(PasoPorSalida);
        Diario.getInstance().ocurreEvento("Jugador " + getNombre() + " pasa por la casilla de salida");
        return true;
    }
    
    boolean pagaImpuesto(float cantidad){
        boolean done=false;
        if(!encarcelado){
            done=paga(cantidad);    
        }
        
        return done;
    }
    
    boolean pagaAlquiler(float cantidad){
        boolean done=false;
        if(!encarcelado){
            paga(cantidad);
            done=true;
        }
        
        return done;
    }
    
    boolean paga(float cantidad){
        Diario.getInstance().ocurreEvento("Jugador " + getNombre() + " paga " + cantidad + " euros");
        return modificarSaldo(cantidad * (-1));
    }
    
    boolean obtenerSalvoconducto(SorpresaSalirCarcel salvoconducto){
        boolean done=false;
        if(!encarcelado){
            this.salvoconducto=salvoconducto;
            done=true;
        }
        return done;
    }
    
    boolean moverACasilla(int numCasilla){
        boolean done=false;
        if(!encarcelado){
            numCasillaActual=numCasilla;
            puedeComprar=false;
            Diario.getInstance().ocurreEvento("Jugador se mueve a la casilla " + numCasilla);
            done=true;
        }
        
        return done;
    }
    
    boolean modificarSaldo(float cantidad){
        saldo += cantidad;
        
        return true;
    }
    
    protected Jugador(Jugador otro){
        this.nombre=otro.nombre;
        this.saldo=otro.saldo;
        this.arrayPropiedades=otro.arrayPropiedades;
        this.encarcelado=otro.encarcelado;
        this.numCasillaActual=otro.numCasillaActual;
        this.puedeComprar=otro.puedeComprar;
        this.salvoconducto=otro.salvoconducto;
    }
    
    Jugador(String nombre){
        this.nombre=nombre;
        this.encarcelado=false;
        this.numCasillaActual=0;
        this.puedeComprar=true;
        this.saldo=SaldoInicial;
        this.salvoconducto = null;
        this.arrayPropiedades=new ArrayList<>();
        this.especulador=false;
    }
    
    public boolean isEncarcelado(){
        return encarcelado;
    }
    
    boolean hipotecar(int ip){
        boolean result = false;
        if (!isEncarcelado()){
            if (existeLaPropiedad(ip)){
                TituloPropiedad propiedad = arrayPropiedades.get(ip);
                result = propiedad.hipotecar(this);
                if (result){
                    Diario.getInstance().ocurreEvento("El jugador " + getNombre() + " hipoteca la propiedad " + propiedad.toString());
                }
            }
        }
        return result;
    }
    
    public float getSaldo(){
        return saldo;
    }
    
    boolean getPuedeComprar(){
        return puedeComprar;
    }
    
    public ArrayList<TituloPropiedad> getPropiedades(){
        ArrayList<TituloPropiedad> aux = new ArrayList<>();
        
        if (arrayPropiedades.size() == 1){
            aux.add(arrayPropiedades.get(0));
        }
        else{
            for (int i=0 ; i<arrayPropiedades.size(); i++){
                aux.add(arrayPropiedades.get(i));
            }
        }
        
        return aux;
    }

    public ArrayList<String> getPropiedadesString(){
        
        ArrayList<String> aux = new ArrayList<>();
        
        if (arrayPropiedades.size() == 1){
            aux.add(arrayPropiedades.get(0).getNombre());
        }
        else{
            for (int i=0 ; i<arrayPropiedades.size() ; i++){
                aux.add(arrayPropiedades.get(i).getNombre());
            }
        }
        
        return aux;
    }
    
    private float getPremioPasoSalida(){
        return PasoPorSalida;
    }
    
    private float getPrecioLibertad(){
        return PrecioLibertad;
    }
    
    int getNumCasillaActual(){
        return numCasillaActual;
    }
    
    public String getNombre(){
        return nombre;
    }
    
    int getHotelesMax(){
        return HotelesMax;
    }
    
    int getCasasPorHotel(){
        return CasasPorHotel;
    }
    
    int getCasasMax(){
        return CasasMax;
    }
    
    boolean existeLaPropiedad(int ip){
        boolean done=false;
        if(tieneAlgoQueGestionar()){
            done = (arrayPropiedades.size() >= ip);
        }
        
        return done;
    }
    
    boolean encarcelar(int numCasillaCarcel){
        if(debeSerEncarcelado()){
            moverACasilla(numCasillaCarcel);
            encarcelado=true;
            Diario.getInstance().ocurreEvento("Jugador encarcelado");
        }
        
        return encarcelado;
    }
    
    boolean enBancarrota(){
        return saldo < 0;
    }
    
    protected boolean debeSerEncarcelado(){
        boolean carcel=false;
        if(!encarcelado){
            if(!tieneSalvoconducto()){
                carcel=true;
            }
            else{
                perderSalvoconducto();
                Diario.getInstance().ocurreEvento("Jugador pierde salvoconducto");
            }
        }
        
        return carcel;
    }
    
    boolean construirHotel(int ip){
        boolean result=false;
        if(!isEncarcelado()){
            if(existeLaPropiedad(ip)){
                TituloPropiedad propiedad = arrayPropiedades.get(ip);
                if(puedoEdificarHotel(propiedad)){
                    result=propiedad.construirHotel(this);
                    propiedad.derruirCasas(CasasPorHotel, this);
                    Diario.getInstance().ocurreEvento("El jugador " + nombre + " construye un hotel en la propiedad " + propiedad.toString());
                }
            }
        }
        
        return result;
    }
    
    
    boolean construirCasa(int ip){
        boolean result=false;
        if(!isEncarcelado()){
            if(existeLaPropiedad(ip)){
                TituloPropiedad propiedad= arrayPropiedades.get(ip);
                if(puedoEdificarCasa(propiedad)){
                    result=propiedad.construirCasa(this);
                    Diario.getInstance().ocurreEvento("El jugador " + nombre + " construye una casa en la propiedad " + propiedad.toString());
                }
            }
        }

        return result;
    }
    
    
    boolean comprar(TituloPropiedad titulo){
        boolean result=false;
        if(!isEncarcelado()){
            if(puedeComprarCasilla()){
                float precio=titulo.getPrecioCompra();
                if(puedoGastar(precio)){
                    result=titulo.comprar(this);
                    if(result){
                        arrayPropiedades.add(titulo);
                        Diario.getInstance().ocurreEvento("El jugador " + nombre + " compra la propiedad " + titulo.toString());
                    }
                    puedeComprar=false;
                }
            }
        }
        
        return result;
    }
    
    

    public int compareTo(Jugador otro){
        Float aux=otro.getSaldo();
        Float thisSaldo=saldo;
        int resultado = thisSaldo.compareTo(aux);
        return resultado;
    }
    
    int cantidadCasasHoteles(){
        int cantidad=0;
        for(int i=0;i<arrayPropiedades.size();i++){
            cantidad += arrayPropiedades.get(i).cantidadCasasHoteles();
        }
        
        return cantidad;
    }
    
    boolean cancelarHipoteca(int ip){
        boolean result=false;
        if(!isEncarcelado()){
            if(existeLaPropiedad(ip)){
                TituloPropiedad propiedad=arrayPropiedades.get(ip);
                float cantidad = propiedad.getImporteCancelarHipoteca();
                if(puedoGastar(cantidad)){
                    result=propiedad.cancelarHipoteca(this);
                    
                    if(result){
                        Diario.getInstance().ocurreEvento("El jugador " + nombre + " cancela la hipoteca de la propiedad " + ip);
                    }
                }
            }
        }
        return result;
    }
    
    public boolean isEspeculador(){
        return especulador;
    }
}

File: CivitasRuby/pkg/CivitasRuby-0.0.1/lib/jugador.rb
# encoding:utf-8 
require_relative 'titulo_propiedad.rb'
require_relative 'sorpresa.rb'
require_relative 'diario.rb'
require_relative 'civitas_juego.rb'
require_relative 'sorpresa_salir_carcel'
require_relative 'casilla'

module Civitas
  class Jugador
    
    attr_reader :casas_max, :casas_por_hotel, :hoteles_max, :num_casilla_actual, :precio_libertad, :paso_por_salida, :puede_comprar, :propiedades, :saldo, :salvoconducto
    
    attr_writer :puede_comprar
    
    @@casas_max = 4
    @@casas_por_hotel = 4
    @@hoteles_max = 4
    @@paso_por_salida = 1000
    @@precio_libertad = 200
    @@saldo_inicial = 7500
    
    def initialize(n)
      @nombre = n
      @encarcelado = false
      @num_casilla_actual = 0
      @puede_comprar = true
      @saldo = @@saldo_inicial
      @salvoconducto = nil
      @propiedades = []
    end
    
    def copia(cop)
      @nombre=cop.get_nombre
      @saldo=cop.saldo
      @propiedades=cop.propiedades
      @encarcelado=cop.is_encarcelado
      @num_casilla_actual=cop.num_casilla_actual
      @puede_comprar=cop.puede_comprar
      @salvoconducto=cop.salvoconducto
    end
    
    def get_nombre()
      return @nombre
    end
    
    def cantidad_casas_hoteles()
      cant = 0
      for i in 0..(@propiedades.length()-1)
        cant += @propiedades[i].cantidad_casas_hoteles()
      end
      return cant
    end
    
    def tiene_salvoconducto()
      return (@salvoconducto != nil)
    end
    
    def is_encarcelado()
      return @encarcelado
    end
    
    def obtener_salvoconducto(sorpresa) #salvoconducto es tipo sorpresa
      done = false
      if (!@encarcelado)
        @salvoconducto = sorpresa
        done = true          
      end
      return done
    end
    
    def perder_salvoconducto()
      Diario.instance.ocurre_evento("Jugador pierde salvoconducto")
      @salvoconducto.usada()
      @salvoconducto = nil
    end
    
    def get_premio_paso_salida()
      return @@paso_por_salida
    end
    
    def debe_ser_encarcelado()
      carcel = false
      if (!@encarcelado)
        if (!tiene_salvoconducto())
          carcel = true
        else
          perder_salvoconducto()
        end
      end
      return carcel
    end
    
    def mover_a_casilla(num_casilla)
      done = false
      if (!@encarcelado)
        @num_casilla_actual = num_casilla
        @puede_comprar = false
        Diario.instance.ocurre_evento("Jugador se mueve a la casilla #{num_casilla}")
        done = true
      end
      return done
    end
    
    def encarcelar(num_casilla_carcel)
      if (debe_ser_encarcelado())
        mover_a_casilla(num_casilla_carcel)
        @encarcelado = true   
        Diario.instance.ocurre_evento("jugador encarcelado")
      end
      return @encarcelado
    end
    
    def puede_comprar_casilla()
      puede = false
      if (!@encarcelado)
        if(@puede_comprar)
          puede = true
        end
      end
      return puede
    end
    
    def modifica_saldo(cant)
      @saldo += cant
      
      return true
    end
    
    def paga(cant)
      return modifica_saldo(cant * (-1))
    end    
    
    def paga_impuesto(cant)
      done = false
      if (!@encarcelado)
        paga(cant)
        done = true
      end
      return done
    end
    
    def paga_alquiler(cant)
      done = false
      if (!@encarcelado)
        paga(cant)
        done = true
      end
      return done
    end
    
    def recibe (cant)
      done = false
      if (!@encarcelado)
        modifica_saldo(cant)
        done = true
      end
      
      return done
    end
    
    def puedo_gastar(precio)
      puedo = @saldo >= precio
      return puedo
    end
    
    def get_propiedades()
      return @propiedades
    end
    
    def tiene_algo_que_gestionar()
      return @propiedades.length > 0   
    end
    
    def existe_la_propiedad(ip)
      aux=ip
      done = false
        if (tiene_algo_que_gestionar())
          if(aux < (@propiedades.length))
            done = true
          end
        end
        return done
    end
    
    def get_propiedades
      aux = []
      
      if (@propiedades.length) == 1
        aux << @propiedades[0].to_string
      else
        for i in 0..(@propiedades.length-1)
          aux << @propiedades[i].to_string
        end
      end
      
      return aux
    end    
    
    def vender(ip)
      done = false
      if (!@encarcelado)
        if (existe_la_propiedad(ip))
          done2 = @propiedades[ip].vender(self)
          if done2
            @propiedades.delete_at(ip)
            Diario.instance.ocurre_evento("Se vende propiedad #{ip}")
            done = true
          end
        end
      end
      return done
    end
    
    def puede_salir_carcel_pagando()
      return puedo_gastar(@@precio_libertad)
    end

    def salir_carcel_pagando()
      done = false
      if(@encarcelado && puede_salir_carcel_pagando())
        paga(@@precio_libertad)
        @encarcelado = false
        done = true
        Diario.instance.ocurre_evento("Jugador sale de la carcel pagando")
      end
      return done
    end
    
    def salir_carcel_tirando()
      done = false
      if(@encarcelado && Dado.instance.salgo_de_la_carcel())
        @encarcelado = false
        done = true
        Diario.instance.ocurre_evento("Jugador sale de la carcel tirando")
      end
      return done
    end
    
    def pasa_por_salida()
      modifica_saldo(@@paso_por_salida)
      Diario.instance.ocurre_evento("Jugador pasa por la casilla de salida")
      return true
    end
    
    def compare_to(otro)
      return @saldo - otro.saldo
    end
    
    def get_saldo
      return @saldo
    end
    
    def en_bancarrota
      result = (@saldo < 0)
      return result
    end
    
def puedo_edificar_hotel(propiedad)
      done=false
      precio = propiedad.precio_edificar
      if puedo_gastar(precio)
        if propiedad.num_hoteles<@@hoteles_max
           if propiedad.num_casas>=@@casas_por_hotel
             done = true
           end
        end
      end
      return done
end
    
def puedo_edificar_casa(propiedad)
      done=false
      precio = propiedad.precio_edificar
      if puedo_gastar(precio)
        if propiedad.num_casas<@@casas_max
          done = true
        end
      end
   return done
end
    
    def cancelar_hipoteca(ip)
      result = false
      if (!is_encarcelado())
        if (existe_la_propiedad(ip))
          propiedad = @propiedades[ip]
          cantidad = propiedad.get_importe_cancelar_hipoteca()
          puedo_gastar = self.puedo_gastar(cantidad)
          if puedo_gastar
            result = propiedad.cancelar_hipoteca(self)
            if result
              Civitas::Diario.instance.ocurre_evento("El jugador #{@nombre} cancela la hipoteca de la propiedad #{ip}")
            end
          end
        end
      end
      return result
    end
    
    def comprar(titulo)
      result = false
      if (!is_encarcelado())
        if (puede_comprar_casilla())
          precio= titulo.precio_compra;
          if (puedo_gastar(precio))
            result = titulo.comprar(self)
            if result
              @propiedades << titulo
              Civitas::Diario.instance.ocurre_evento("El jugador #{get_nombre} compra la propiedad " + titulo.to_string())
            end
            @puede_comprar= false
          end
        end
      end
      return result
    end
    
    def construir_hotel(ip)
      result = false
      if !is_encarcelado()
        if existe_la_propiedad(ip)
          propiedad = @propiedades[ip]
          puedo_edificar_hotel = puedo_edificar_hotel(propiedad)
          if puedo_edificar_hotel
            result = propiedad.construir_hotel(self)
            casas_por_hotel = @@casas_por_hotel
            propiedad.derruir_casas(casas_por_hotel,self)
            Civitas::Diario.instance.ocurre_evento("El jugador #{get_nombre} construye un hotel en la propiedad " + propiedad.to_string())
          end
          
        end
      end
      return result
    end
    
    def construir_casa(ip)
      result = false
      if !is_encarcelado()
        if existe_la_propiedad(ip)
          propiedad = @propiedades[ip]
          puedo_edificar_casa = puedo_edificar_casa(propiedad)
          if puedo_edificar_casa
            result = propiedad.construir_casa(self)
            Civitas::Diario.instance.ocurre_evento("El jugador #{get_nombre} construye una casa en la propiedad " + propiedad.to_string())
          end
          
        end
      end
      return result
    end
    
    def hipotecar(ip)
    result = false
      if (!is_encarcelado())
          if existe_la_propiedad(ip)
            propiedad = @propiedades[ip]
            result = propiedad.hipotecar(self)
            if result
              Civitas::Diario.instance.ocurre_evento("El jugador #{get_nombre} hipoteca la propiedad " + propiedad.to_string())
            end
          end
      end
      return result
    end
    
    def to_string()
      String::new(mensaje="******************************\n")
      mensaje += @nombre + "\n******************************\n\n" + "ENCARCELADO = " + @encarcelado.to_s 
      mensaje += "\n Numero Casilla Actual: #{@num_casilla_actual}         Saldo: #{@saldo}€"
      mensaje += "\n Puede comprar= " + @puede_comprar.to_s
      mensaje += "\n CasasMax: #{@@casas_max}      Casas por hotel: #{@@casas_por_hotel}      Hoteles max: #{@@hoteles_max}"
      mensaje += "\n Paso por salida = #{@@paso_por_salida}€, Precio Libertad = #{@@precio_libertad}€, Saldo inicial = #{@@saldo_inicial}"
      
      return mensaje
    end
    
    
  end
end